scanner/src/main/kotlin/LocalScanner.kt

```diff
@@ -131,42 +137,106 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
+                                // Remove the now unneeded reference to rawResult here to allow garbage collection to
+                                // clean it up.
+                                it.copy(rawResult = null)
+                            }
+                        } catch (e: ScanException) {
+                            e.showStackTrace()
+
+                            log.error {
+                                "Could not scan '${pkg.id.toCoordinates()}' $packageIndex: " +
+                                        e.collectMessagesAsString()
+                            }
+
+                            val now = Instant.now()
+                            listOf(ScanResult(
+                                    provenance = Provenance(),
+                                    scanner = scannerDetails,
+                                    summary = ScanSummary(
+                                            startTime = now,
+                                            endTime = now,
+                                            fileCount = 0,
+                                            licenseFindings = sortedSetOf(),
+                                            errors = listOf(OrtIssue(
+                                                    source = scannerName,
+                                                    message = e.collectMessagesAsString()
+                                            ))
+                                    ),
+                                    rawResult = EMPTY_JSON_NODE)
+                            )
+                        }
+
+                        Pair(pkg, result)
+                    }
+                }.associate { it.await() }
             }
-
-            Pair(pkg, result)
+        } finally {
+            cacheDispatcher.close()
+            scanDispatcher.close()
         }
     }
 
+    private fun getResultsFile(scannerDetails: ScannerDetails, pkg: Package, outputDirectory: File): File {
+        val scanResultsForPackageDirectory = File(outputDirectory, pkg.id.toPath()).apply { safeMkdirs() }
+        return File(scanResultsForPackageDirectory, "scan-results_${scannerDetails.name}.$resultFileExt")
+    }
+
+    private fun readFromCache(scannerDetails: ScannerDetails, pkg: Package, outputDirectory: File): List<ScanResult> {
+        val resultsFile = getResultsFile(scannerDetails, pkg, outputDirectory)
+
+        val storedResults = ScanResultsStorage.read(pkg, scannerDetails)
+
+        return if (storedResults.results.isNotEmpty()) {
+            // Some external tools rely on the raw results filer to be written to the scan results directory, so write
+            // the first stored result to resultsFile. This feature will be removed when the reporter tool becomes
+            // available.
+            resultsFile.mapper().writeValue(resultsFile, storedResults.results.first().rawResult)
+            storedResults.results
+        } else emptyList()
```

Comment:
In this `else` branch `storedResults.results` is empty, so can't we generally return `storedResults.results`?

Reply:
Done.